core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

```diff
@@ -160,23 +160,80 @@ class OptimizeExecutor(
 
   private val isMultiDimClustering = zOrderByColumns.nonEmpty
 
-  def optimize(): Seq[Row] = {
+  def optimize(isAutoCompact: Boolean = false, targetFiles: Seq[AddFile] = Nil): Seq[Row] = {
     recordDeltaOperation(txn.deltaLog, "delta.optimize") {
       val minFileSize = sparkSession.sessionState.conf.getConf(
         DeltaSQLConf.DELTA_OPTIMIZE_MIN_FILE_SIZE)
-      val maxFileSize = sparkSession.sessionState.conf.getConf(
-        DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE)
+      val maxFileSize = if (isAutoCompact) {
+        sparkSession.sessionState.conf.getConf(DeltaSQLConf.AUTO_COMPACT_MAX_FILE_SIZE)
+      } else {
+        sparkSession.sessionState.conf.getConf(DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE)
+      }
       require(minFileSize > 0, "minFileSize must be > 0")
       require(maxFileSize > 0, "maxFileSize must be > 0")
 
-      val candidateFiles = txn.filterFiles(partitionPredicate)
+      val minNumFilesInDir = if (isAutoCompact) {
+        val minNumFiles =
+          sparkSession.sessionState.conf.getConf(DeltaSQLConf.AUTO_COMPACT_MIN_NUM_FILES)
+        require(minNumFiles > 0, "minNumFiles must be > 0")
+        minNumFiles
+      } else if (isMultiDimClustering) {
+        1
+      } else {
+        // compaction
+        2
```

Comment:
Introduce constants (perhaps `sealed trait` with `case object`s?) for this magic numbers

Reply:
Could you provide the definitions? not sure what would be better. 
Added comments.

Reply:
Let's start with a couple of `val`s to give some meaning to these numbers first, e.g.

```
val MULTI_DIM_CLUSTERING=1
val COMPACTION_MORE_FILES = 2
```

(not sure these names are correct but that's the idea)

Reply:
@jaceklaskowski I refactored the part as the function is already bit long. 
Could you review the PR again? Thanks :)